### PR TITLE
feat: add new room-invite event

### PR DESCRIPTION
**Pull request by windmemory**
refer to RFC https://github.com/Chatie/wechaty/issues/1492

### CODE REVIEW

File: src/puppet.ts [hunk 1]
```diff
@@ -257,17 +258,18 @@ export abstract class Puppet extends EventEmitter {
...
+  public emit (event: 'error',       error: Error)                                                          : boolean
+  public emit (event: 'friendship',  friendshipId: string)                                                  : boolean
+  public emit (event: 'login',       contactId: string)                                                     : boolean
+  public emit (event: 'logout',      contactId: string)                                                     : boolean
+  public emit (event: 'message',     messageId: string)                                                     : boolean
+  public emit (event: 'reset',       reason: string)                                                        : boolean
+  public emit (event: 'room-join',   roomId: string, inviteeIdList: string[],  inviterId: string)           : boolean
+  public emit (event: 'room-leave',  roomId: string, leaverIdList: string[], remover?: string)              : boolean
+  public emit (event: 'room-topic',  roomId: string, newTopic: string, oldTopic: string, changerId: string) : boolean
+  public emit (event: 'room-invite', roomInvitation: RoomInvitation)                                        : boolean
```
Comment (zixia) [inline src/puppet.ts]: Please leave other `emit()` lines to be un-modified in your PR, to make it clear.

File: src/puppet.ts [hunk 2]
```diff
@@ -289,21 +291,22 @@ export abstract class Puppet extends EventEmitter {
...
+  public on (event: 'login',        listener: (contactId: string) => void)                                                     : this
+  public on (event: 'logout',       listener: (contactId: string) => void)                                                     : this
+  public on (event: 'message',      listener: (messageId: string) => void)                                                     : this
+  public on (event: 'reset',        listener: (reason: string) => void)                                                        : this
+  public on (event: 'room-join',    listener: (roomId: string, inviteeIdList: string[], inviterId:  string) => void)           : this
+  public on (event: 'room-leave',   listener: (roomId: string, leaverIdList : string[], removerId?: string) => void)           : this
+  public on (event: 'room-topic',   listener: (roomId: string, newTopic: string, oldTopic: string, changerId: string) => void) : this
+  public on (event: 'room-invite',  listener: (roomInvitation: RoomInvitation) => void)                                        : this
```
Comment (zixia) [inline src/puppet.ts]: Only keep this line to be modified.

File: src/schemas/puppet.ts [hunk 3]
```diff
@@ -36,23 +41,24 @@ export interface PuppetRoomTopicEvent {
...
-  message     : 'document can be writen at here',
-  'room-join' : 'document can be writen at here',
-  'room-leave': 'document can be writen at here',
-  'room-topic': 'document can be writen at here',
-  scan        : 'document can be writen at here',
+  friendship   : 'document can be written at here',
+  login        : 'document can be written at here',
+  logout       : 'document can be written at here',
+  message      : 'document can be written at here',
+  'room-invite': 'document can be written at here',
```
Comment (zixia) [inline src/schemas/puppet.ts]: Only keep this line.

We can ignore the code style format in this PR, for better readability of the diff.

File: src/schemas/puppet.ts [hunk 4]
```diff
@@ -36,23 +41,24 @@ export interface PuppetRoomTopicEvent {
...
 
 export const PUPPET_EVENT_DICT = {
   ...CHAT_EVENT_DICT,
-  dong      : 'document can be writen at here',
-  error     : 'document can be writen at here',
+  dong      : 'document can be written at here',
+  error     : 'document can be written at here',
   reset     : 'push to reset!',
-  watchdog  : 'document can be writen at here',
+  watchdog  : 'document can be written at here',
```
Comment (zixia) [inline src/schemas/puppet.ts]: It's better to do `writen` -> `written` in another PR instead of mixing them at here.

**Comment by windmemory** [comment]
I will not change the code style/spell correction next time, but for this PR, I think reverting the styles and spell correction back to previous one is just wasting time. 

The reason I submit PR separately is that, firstly doing `npm link` in three repository is painful, more importantly, `wechaty-puppet` is an abstract of each puppet, I think it would be better to confirm the design on puppet layer first. So I submit this PR.

If I have the implementation done, but the interface in puppet is not good, then I need to change the code from bottom to top, that is wasting time. So please let me know if you think the interface here is good and we can confirm with this interface, then I will submit more PR in the other two repository.

**Comment by zixia** [comment]
Sorry but I'm afraid that you have to be used to work with `npm link` because that's the only way when we develop multi-dependencies modules.

And I'm afraid that I can not tell you whether the interface here is good or not, because as I said: all of those is a whole part of one thing, we have too few information about the design when we just get started.

So my suggestion is to be prepared to refactor your code and start coding with the 3 modules with the help of `npm link`.

If you still feel that the time we will spend is a waste, I hope you can agree with me that that's the trade-off for the decouple & systematic thinking.

**Comment by windmemory** [comment]
> And I'm afraid that I can not tell you whether the interface here is good or not, because as I said: all of those is a whole part of one thing, we have too few information about the design when we just get started.

If all of those three repository is a whole part, then it doesn't make sense to separate those, right? Now you separate those three repository proof that you believe they are serving as different component in the `wechaty` eco-system. So they must have their own scope, otherwise, they should be all in one repository. 

From my understanding, `wechaty` in charge of linking outside developers with our inside implementation, it converts our internal impl to an easy to use API. `wechaty-puppet` is behaving as the connector between wechaty and underline protocol implementations, so it defines what kind of interfaces is expected from `wechaty` perspective, and it doesn't care about how the underline protocol implementation layer actually do. Finally the `wechaty-puppet-padchat` or others like `wecahty-puppet-puppteer` or `wechaty-puppet-wechat4u`, they are doing the job linking outside protocol, convert the data/messages to the format that `puppet` is able to process. 

So I would like to discuss that what is expecting from `wechaty` layer for the new event. And what's the data is expected, is the current structure good for `wechaty`? 

If you need to see the actual message from `padchat`, I could open a raw PR to let you know the data format etc. But I think the high level discussion in this repo is not blocked.

**Comment by zixia** [comment]
Thanks for the reply.

I agree with you partly, but let's postpone to discuss it later before you have a working prototype.

Please write whatever you want to make it just workable, and let's discuss it based on your workable code.

**Comment by zixia** [comment]
After I study your workable prototype,  I had made some design changes based on your PR.

Please see my commits for details, and thank you for this PR!

**Pull request merged**